src/runtime_src/core/common/api/xrt_xclbin.cpp

```diff
@@ -38,126 +41,919 @@
 
 namespace xrt {
 
-// class xclbin_impl - class for xclbin objects
-//
-// Life time of xclbin are managed through shared pointers.
-// A buffer is freed when last references is released.
-class xclbin_impl
+/**
+ * class xclbin::mem_impl - xrt::xclbin::mem_impl provides the implementation
+ *
+ *   for a device memory object.
+ *
+ */
+class xclbin::mem_impl
 {
-protected:
-  std::vector<char> m_axlf;
-  const axlf* m_top;
+public:
+  mem_impl(const struct mem_data& memdata, const int32_t memidx);
 
-  void init_axlf_handle()
+  const std::string&
+  get_name() const{
+    return m_name;
+  }
+  
+  uint64_t
+  get_base_address() const
   {
-    const axlf* tmp = reinterpret_cast<const axlf*>(m_axlf.data());
-    if (strncmp(tmp->m_magic, "xclbin2", 7)) // Future: Do not hardcode "xclbin2"
-      throw std::runtime_error("Invalid xclbin");
-    m_top = tmp;
+    return m_base_address;
+  }
+  
+  uint64_t
+  get_size() const
+  {
+    return m_size;
   }
 
-public:
-  xclbin_impl(const std::vector<char>& data)
+  uint8_t
+  get_used() const
   {
-    m_axlf = data;
-    // Set pointer of type axlf*, and check magic string
-    init_axlf_handle();
+    return m_used;
   }
 
-  xclbin_impl(const std::string& filename)
+  uint8_t
+  get_type() const
   {
-    if (filename.empty())
-      throw std::runtime_error("No XCLBIN specified");
+    return m_type;
+  }
+
+  int32_t
+  get_index() const
+  {
+    return m_index;
+  }
 
-    // Load the file into data
-    std::ifstream stream(filename);
-    stream.seekg(0, stream.end);
-    size_t size = stream.tellg();
-    stream.seekg(0, stream.beg);
-    m_axlf.resize(size);
-    stream.read(m_axlf.data(), size);
+protected:
+  std::string m_name;
+  uint64_t m_base_address;
+  uint64_t m_size;
+  uint8_t m_used;
+  uint8_t m_type;
+  int32_t m_index;
+};
+  
+/**
```

Comment:
For consistently with out APIs please move the definition of the APIs from xrt_xclbin.h to after all the internal impl classes.